src/include/storage/garbage_collector.h

```diff
@@ -0,0 +1,180 @@
+   * a while loop to handle contention from running transactions (basically restart the process if needed).
+   * @param txn pointer to the transaction that created this DeltaRecord
+   * @param undo_record DeltaRecord to be unlinked
+   */
+  void UnlinkDeltaRecord(transaction::TransactionContext *const txn, const DeltaRecord &undo_record) {
+    PELOTON_ASSERT(txn->TxnId() == undo_record.Timestamp().load(), "This undo_record does not belong to this txn.");
+    DataTable *const table = undo_record.Table();
+    const TupleSlot slot = undo_record.Slot();
+    const TupleAccessStrategy &accessor = table->accessor_;
+
+    DeltaRecord *version_ptr;
+    do {
+      version_ptr = table->AtomicallyReadVersionPtr(slot, accessor);
+      PELOTON_ASSERT(version_ptr != nullptr, "GC should not be trying to unlink in an empty version chain.");
+
+      if (version_ptr->Timestamp().load() == txn->TxnId()) {
+        // Our DeltaRecord is the first in the chain, handle contention on the write lock with CAS
+        if (table->CompareAndSwapVersionPtr(slot, accessor, version_ptr, version_ptr->Next())) break;
+        // Someone swooped the VersionPointer while we were trying to swap it (aka took the write lock)
+        version_ptr = table->AtomicallyReadVersionPtr(slot, accessor);
+      }
+      // a version chain is guaranteed to not change when not at the head (assuming single-threaded GC), so we are safe
+      // to traverse and update pointers without CAS
+      DeltaRecord *curr = version_ptr;
+      DeltaRecord *next = curr->Next();
+
+      // traverse until we hit the DeltaRecord that we want to unlink
+      while (next != nullptr && next->Timestamp().load() != txn->TxnId()) {
+        curr = next;
+        next = curr->Next();
+      }
+      // we're in position with next being the DeltaRecord to be unlinked
+      if (next != nullptr && next->Timestamp().load() == txn->TxnId()) {
+        curr->Next().store(next->Next().load());
+        break;
+      }
+      // If that process didn't work (interleaved abort) then try again
+    } while (true);
+  }
+
+  /**
+   * Process the unlink queue
+   * @return number of txns unlinked (not DeltaRecords) for debugging/testing
+   */
+  uint32_t Unlink() {
+    const timestamp_t oldest_txn = txn_manager_->OldestTransactionStartTime();
+    transaction::TransactionContext *txn = nullptr;
+
+    // Get the completed transactions from the TransactionManager
+    std::queue<transaction::TransactionContext *> from_txn_manager = txn_manager_->CompletedTransactions();
+    if (!txns_to_unlink_.empty()) {
+      // Append to our non-empty unlink queue
+      while (!from_txn_manager.empty()) {
+        txn = from_txn_manager.front();
+        from_txn_manager.pop();
+        txns_to_unlink_.push(txn);
+      }
+    } else {
+      // Overwrite our empty unlink queue
+      txns_to_unlink_ = from_txn_manager;
```

Comment:
nit: This probably copies anyway, so this if branch can probably go away. (simply iterate through the `from_txn_manager` and add them to `txns_to_unlink`)